File: src/Renderer.cpp
#include "Renderer.h"
#include "Palettes.h"
#include <string>
#include <stdio.h>
#include <stdlib.h>
#include <ctime>

std::string _PATH = "../_frames/";

void
Renderer::_memory_render(uint8_t* lut_values) {
    // 3 bytes per pixel.
    const uint8_t bytes_per_pixel = 3;
    // FIXME
    // Not actual pixels, theyre LUT values.
    for (int y = 0; y < 240; y++) {
        // Go row by row.
        for (int x = 0; x < 256; x++) {
            // Each column in the row.
            uint64_t input_idx = x + y * 256;
            uint8_t pixel_lookup = lut_values[input_idx];
            // Draw a debug grid over the output.
            // if (x % 8 == 0 || y % 8 == 0) {
            //     // Debug nametable grid with yellow.
            //     pixel_lookup = 0x37;
            // }
            // if (x % 0x20 == 0 || y % 0x20 == 0) {
            //     // Debug attribute grid with red.
            //     pixel_lookup = 0x16;
            // }
            // if (x % 0x10 == 0 || y % 0x10 == 0) {
            //     // Debug attributes quadrants with pink.
            //     pixel_lookup = 0x25;
            // }

            uint64_t output_idx = x * bytes_per_pixel + y * 256 * bytes_per_pixel;
            pixels[output_idx] = LUT[pixel_lookup][0];
            pixels[output_idx + 1] = LUT[pixel_lookup][1];
            pixels[output_idx + 2] = LUT[pixel_lookup][2];
        }
    }
}

void
Renderer::_file_render(uint8_t* lut_values) {
    std::string filepath = "-frame_" + std::to_string(_frames) + ".ppm";
    if (file_prefix != "") {
        filepath = file_prefix + filepath;
    } else {
        filepath = std::to_string(std::time(nullptr)) + filepath;
    }
    filepath = _PATH + filepath;

    // printf("%s\n", filepath.c_str());
    FILE* f = fopen(filepath.c_str(), "a+");
    if (NULL == f) {
        printf("Failed to render to file at path %s\n", filepath.c_str());
        return;
    }

    fprintf(f, "P3\n");
    fprintf(f, "256 240\n");
    fprintf(f, "255\n");
    for (int y = 0; y < 240; y++) {
        // Go row by row.
        for (int x = 0; x < 256; x++) {
            // Each column in the row.
            uint64_t idx = x + y * 256;
            uint8_t pixel_lookup = lut_values[idx];

            // Draw a debug grid over the output.
            // if (x % 8 == 0 || y % 8 == 0) {
            //     // Debug nametable grid with yellow.
            //     pixel_lookup = 0x37;
            // }
            // if (x % 0x20 == 0 || y % 0x20 == 0) {
            //     // Debug attribute grid with red.
            //     pixel_lookup = 0x16;
            // }
            // if (x % 0x10 == 0 || y % 0x10 == 0) {
            //     // Debug attributes quadrants with pink.
            //     pixel_lookup = 0x25;
            // }
            fprintf(f, "%d %d %d ",
                LUT[pixel_lookup][0],
                LUT[pixel_lookup][1],
                LUT[pixel_lookup][2]
            );
        }
        fprintf(f, "\n");
    }
    fclose(f);
}

void
Renderer::render(uint8_t* pixels) {
    _frames++;
    // Did 1 second elapse?
    time_t now = time(nullptr);
    double delta_seconds = difftime(now, _time);
    // TODO
    // Find a better way to compute FPS.
    if (delta_seconds >= 1) {
        // Around 1 second has elapsed, how many frames were
        // made in that time?
        uint64_t fps = _frames - _last_frames;
        _time = now;
        _last_frames = _frames;
        // TODO
        // Temp.
        printf("FPS: %llu %lf\n", fps, delta_seconds);
    }
    switch (render_type) {
        case RenderType::MEMORY:
            return _memory_render(pixels);
        case RenderType::FILE:
            return _file_render(pixels);
        case RenderType::NONE:
            printf("frame %llu\n", _frames);
            return;
    }
}

File: py/src/renderer.py
import sys
from datetime import datetime
from pathlib import Path
from time import time

from palettes import LUT


_PATH = '_frames/'
# Allows for 'debouncing' frame disk writes.
_FILE_PER_FRAMES = 1

_RENDERER_TYPE = 'file'
# _RENDERER_TYPE = 'memory'

class Renderer():
    """
    Takes the entire pixel buffer for one frame and renders it.
    """

    def __init__(self) -> None:
        self.renderer_type = _RENDERER_TYPE
        self._start_dt = datetime.now().strftime('%Y-%m-%d-%H-%M-%S')
        self._frames = 0
        self._last_frame = time()

    def _file_render(self, pixels: bytearray) -> None:
        if self._frames % _FILE_PER_FRAMES != 0:
            return
        filepath = f'{_PATH}{self._start_dt}-frame_{self._frames}.ppm'
        with open(filepath, 'w') as f:
            f.write('P3\n')
            f.write('256 240\n')
            f.write('255\n')
            for y in range(240):
                # Go row by row.
                for x in range(256):
                    # Each column in the row.
                    idx = x + y * 256
                    pixel_lookup = pixels[idx]
                    try:
                        r, g, b = LUT[pixel_lookup]
                        f.write(f'{r} {g} {b} ')
                    except:
                        f.write(f'0 0 0 ')
                f.write('\n')

    def _terminal_render(self, pixels: bytearray) -> None:
        # Lookup for pixel codes to terminal colours. We'd need
        # a terminal-oriented palette for this that isn't really RGB.
        lut = {
            # Black
            0x0f: 0,
            # Dark blue
            0x12: 26,
            # Pink?
            0x25: 207,
            # Orange
            0x27: 208,
            # Light blue
            0x2c: 45,
            # White
            0x30: 15,
            # Brownish
            0x38: 173,
        }
        print()
        for i, p in enumerate(pixels):
            if p not in lut:
                raise Exception(
                    f"Attempting to output a colour that isn't in the LUT {hex(p)}"
                )
            code = str(lut[p])
            sys.stdout.write(u"\u001b[38;5;" + code + "m" + '■ ')
            sys.stdout.flush()
            if i > 0 and i % 256 == 0:
                sys.stdout.write('\n')
                sys.stdout.flush()
        print()

    def render(self, pixels: bytearray) -> None:
        self._frames += 1
        if self.renderer_type == 'terminal':
            return self._terminal_render(pixels)
        elif self.renderer_type == 'file':
            return self._file_render(pixels)
        elif self.renderer_type == 'memory':
            n = time()
            d = n - self._last_frame
            self._last_frame = n
            print(d)
            return None
